app/apollo/resolvers/group.js

```diff
@@ -310,17 +327,31 @@ const groupResolvers = {
       const user = whoIs(me);
 
       try {
-        logger.info({ req_id, user, org_id, groupUuids, clusterIds }, `${queryName} validating`);
-
-        await validAuth(me, org_id, ACTIONS.MANAGE, TYPES.GROUP, queryName, context);
+        logger.info({req_id, user, org_id, groupUuids, clusterIds}, `${queryName} validating`);
 
         validateString( 'org_id', org_id );
         groupUuids.forEach( value => { validateString( 'groupUuids', value ); } );
         clusterIds.forEach( value => validateString( 'clusterIds', value ) );
 
-        const groups = await models.Group.find({org_id, uuid: {$in: groupUuids}});
-        if (groups.length < 1) {
-          throw new NotFoundError(context.req.t('None of the passed group uuids were found'));
+        let groups = await getAllowedResources(me, org_id, ACTIONS.MANAGE, TYPES.GROUP, queryName, context, null, groupUuids);
+        logger.info({req_id, user, org_id, groupUuids, clusterIds}, `${queryName} retrieved allowed resources`);
+
+        if (groups.length < 1) { throw new NotFoundError(context.req.t('None of the passed group uuids were found')); }
+
+        let allAllowedClusters = false;
+        try {
+          await validAuth(me, org_id, ACTIONS.READ, TYPES.CLUSTER, queryName, context);
+          allAllowedClusters = true;
+        }
+        catch(e){ // If exception thrown, user does NOT have auth to all resources of this type, and code must later filter based on fine grained auth
+        }
+        logger.info({req_id, user, org_id, groupUuids, clusterIds, allAllowedClusters}, `${queryName} validating - allAllowed: ${allAllowedClusters}`);
+
+        let clusters = await commonClusterSearch(models, {org_id}, { limit: 0, skip: 0, startingAfter: null });
+
+        if (!allAllowedClusters){
+          clusters = await filterResourcesToAllowed(me, org_id, ACTIONS.ATTACH, TYPES.CLUSTER, clusters, context);
+          logger.info({req_id, user, org_id}, `${queryName} filtered resources to allowed`);
```

Comment:
If user specified clusters that do not exist or to which they are not authorized, should the call fail?  I.e. if the allowed clusters length is less than the number of clusterIds passed, should it fail?

Similarly for Gropus, if user specified groupUuids for gorups they don't have access to (or dont exist), should it fail?  Currently (and originally) it would only fail if _none_ of the passed groups existed.

Reply:
Yes they should fail. I added in from editClusterGroups:
`        if (groups.length != groupUuids.length) {
          throw new NotFoundError(context.req.t('One or more of the passed group uuids were not found'));
        }`

Reply:
We didn't fail clusters not existing for editClusterGroups and implementation appears to be very similar. Probably leave this part as is even though it is bad implementation.

Reply:
When bad implementation is noted in an area being changed, either fix the bad implementation or at least add a code comment explaining that yes the problem exists and future changes should consider fixing it, e.g.:
```
/*
Note: if the user passes clusters that do not exist or to which the user does not have access, those clusters are treated as if they were not passed.  It would be better throw an exception if the clusters were not found / not authorized -- future enhancement should consider fixing this.
*/
```

**However**, in this case as this PR is adding fine grained authorization checks, it would be much better / more appropriate to actually _fix_ the problem.

Reply:
Unresolving conversation, not fully addressed.

Reply:
Opinion: For assignClusterGroups, Throw error, add unit test. I can't think of a situation where the user would have permission to group specific clusters that do not currently exist and where this function would need to continue. Clusters should be created and then grouped in that case.

Reply:
I agree

Reply:
Resolved.